explorer/interpreter/interpreter.cpp

```diff
@@ -1934,44 +1944,101 @@ auto Interpreter::StepDeclaration() -> ErrorOr<Success> {
-          const auto& class_members = class_dec.members();
-          for (const auto& member : class_members) {
-            if (const auto* var = dyn_cast<VariableDeclaration>(member)) {
-              const auto& type = var->static_type();
-              if (const auto* c_type = dyn_cast<NominalClassType>(&type)) {
-                auto& c_dec = c_type->declaration();
-                if (c_dec.destructor().has_value()) {
-                  Address object = lvalue->address();
-                  Address mem = object.SubobjectAddress(Member(var));
-                  auto v = heap_.Read(mem, source_loc);
-                  act.scope()->TransitState();
-                  return CallDestructor(*c_dec.destructor(), *v);
-                }
-              }
-            }
+      }
+      if (const auto* class_obj =
+              dyn_cast<NominalClassValue>(*destroy_act.value())) {
+        const auto& class_type = cast<NominalClassType>(class_obj->type());
+        const auto& class_dec = class_type.declaration();
+        if (class_dec.destructor().has_value()) {
+          return CallDestructor(*class_dec.destructor(), class_obj);
+        }
+      }
+    }
+  }
+  if (destroy_act.declaration()) {
+    todo_.Pop();
+    return Success();
+  }
+  if (act.pos() > 0) {
+    if (const auto* class_obj =
+            dyn_cast<NominalClassValue>(*destroy_act.value())) {
+      const auto& class_type = cast<NominalClassType>(class_obj->type());
+      const auto& class_dec = class_type.declaration();
+      int index = class_dec.members().size() - act.pos();
+      if (index >= 0 && index < static_cast<int>(class_dec.members().size())) {
+        const auto& member = class_dec.members()[index];
+        if (const auto* var = dyn_cast<VariableDeclaration>(member)) {
+          Address object = destroy_act.lvalue()->address();
+          Address mem = object.SubobjectAddress(Member(var));
+          SourceLocation source_loc("destructor", 1);
+          auto v = heap_.Read(mem, source_loc);
+          return todo_.Spawn(std::make_unique<DestroyAction>(
+              destroy_act.lvalue(), *v, member));
+        }
+        return todo_.Spawn(std::make_unique<DestroyAction>(
+            destroy_act.lvalue(), std::nullopt, member));
+      }
+    }
+    if (const auto* tuple = dyn_cast<TupleValue>(*destroy_act.value())) {
+      if (tuple->elements().size() > 0) {
+        int index = tuple->elements().size() - act.pos() - 1;
+        if (index >= 0) {
+          auto res = DestroyTupleElement(destroy_act.lvalue(), tuple, index);
+          if (res) {
+            return std::move(*res);
           }
         }
-        act.scope()->TransitState();
+      }
+    }
+  }
+  todo_.Pop();
+  return Success();
+}
+
+auto Interpreter::StepCleanUp() -> ErrorOr<Success> {
+  Action& act = todo_.CurrentAction();
+  CleanupAction& cleanup = cast<CleanupAction>(act);
+  if (act.scope()->kind() != RuntimeScope::Kind::Method) {
```

Comment:
It looks like this won't do _any_ cleanup at the top of a method scope, so it skips the `Self` parameter, but also skips all the function parameters, some of which might need to be destroyed. There are also other cases where a local name goes out of scope but the underlying object shouldn't be destroyed, because the name is effectively an alias. So I don't think treating method scopes differently is the way to go here. Instead, we need to do two things:

- Have `CleanupAction` only clean up locals that have _storage_ in the current scope.
- Stop allocating local storage for names like `me` that act as aliases.

To do that, I suggest changing `RuntimeScope::allocations_` to hold the `Value*`s as well as the allocation IDs, so that it has all the information we need to deallocate and destroy each object. Then we have this code loop over `allocations_` instead of `locals_`. To complement that, we can add a `Bind` method to `RuntimeScope`, which takes the same parameters as `Initialize`, but only binds the name to the value, and doesn't allocate storage for it the way `Initialize` does (and doesn't store anything in `allocations_`). Then we can update the logic that handles `me` to use `Bind` instead of `Initialize`.

We'll also need to update a lot of other code to use `Bind` instead of `Initialize`, and we'll probably also need an API for allocating storage without binding a name to it, but I think that can all be a TODO.

Does that make sense?